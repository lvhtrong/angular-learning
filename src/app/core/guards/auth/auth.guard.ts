import { Injectable } from '@angular/core';
import {
  CanActivateChild,
  ActivatedRouteSnapshot,
  RouterStateSnapshot,
  UrlTree,
  Router,
} from '@angular/router';
import { Observable } from 'rxjs';
import { AuthService } from '@core/domains/auth/services/auth.service';

@Injectable({
  providedIn: 'root',
})
export class AuthGuard implements CanActivateChild {
  constructor(private authService: AuthService, private router: Router) {}

  canActivateChild(
    next: ActivatedRouteSnapshot,
    state: RouterStateSnapshot
  ):
    | Observable<boolean | UrlTree>
    | Promise<boolean | UrlTree>
    | boolean
    | UrlTree {
    const isAuthenticated = this.authService.isAuthenticated();
    if (isAuthenticated) {
      return isAuthenticated;
    } else {
      this.router.navigate(['/auth/login'], {
        queryParams: {
          redirectUrl: '',
        },
      });
      return false;
    }
  }
}
