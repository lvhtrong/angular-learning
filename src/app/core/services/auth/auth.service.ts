import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';

@Injectable({
  providedIn: 'root',
})
export class AuthService implements AuthService {
  constructor(private http: HttpClient) {}

  public login({ username, password }: { username: string; password: string }) {
    console.log(username, password);
  }
}
