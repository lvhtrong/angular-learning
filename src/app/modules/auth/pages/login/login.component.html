<div class="container">
  <div class="row flex justify-content-center align-items-center h-100">
    <div class="col-md-6 card">
      <div class="card-body">
        <h1 class="mb-3">Login</h1>
        <div class="form-group">
          <label for="login-email-input">Username</label>
          <input
            type="email"
            class="form-control"
            id="login-email-input"
            [value]="username"
            (input)="onUsernameChange($event.target.value)"
          />
        </div>
        <div class="form-group">
          <label for="login-password-input">Password</label>
          <input
            type="password"
            class="form-control"
            id="login-password-input"
            [value]="password"
            (input)="onPasswordChange($event.target.value)"
          />
        </div>
        <button type="submit" class="btn btn-primary" (click)="onSubmitClick()">
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
