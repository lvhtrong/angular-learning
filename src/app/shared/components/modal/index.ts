import { ModalNotifyComponent } from './notify/notify.component';

export const modals = [ModalNotifyComponent];
