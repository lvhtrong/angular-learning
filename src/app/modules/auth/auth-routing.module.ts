import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { NonAuthGuard } from '@core/guards/non-auth/non-auth.guard';

import { LoginComponent } from './pages/login/login.component';

const routes: Routes = [
  {
    path: 'auth',
    canActivate: [NonAuthGuard],
    children: [
      {
        path: '',
        redirectTo: 'login',
        pathMatch: 'full',
      },
      {
        path: 'login',
        component: LoginComponent,
      },
    ],
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class AuthRoutingModule {}
