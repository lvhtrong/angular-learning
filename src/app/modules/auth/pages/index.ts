import { LoginComponent } from './login/login.component';

export const pages: any[] = [LoginComponent];
