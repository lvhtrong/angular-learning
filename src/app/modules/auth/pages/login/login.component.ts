import { Component, OnInit } from '@angular/core';
import { AuthService } from '@core/domains/auth/services/auth.service';
import { throwError, Observable, of } from 'rxjs';
import { map, catchError, switchMap } from 'rxjs/operators';
import { ActivatedRoute, Router } from '@angular/router';
import { ModalService } from '@core/services/modal/modal.service';

@Component({
  selector: 'app-login',
  templateUrl: './login.component.html',
  styleUrls: ['./login.component.scss'],
})
export class LoginComponent implements OnInit {
  username = '';
  usernameError = null;

  password = '';
  passwordError = null;

  constructor(
    private authService: AuthService,
    private activatedRoute: ActivatedRoute,
    private router: Router,
    private modalService: ModalService
  ) {}

  ngOnInit(): void {}

  onUsernameChange(value: string) {
    this.username = value;
    this.usernameError = null;
  }

  onPasswordChange(value: string) {
    this.password = value;
    this.passwordError = null;
  }

  onSubmitClick() {
    if (!this.username || !this.password) {
      if (!this.username) {
        this.usernameError = 'User Name is required';
      }
      if (!this.password) {
        this.passwordError = 'Password is required';
      }
      return;
    }

    this.authService
      .login({
        username: this.username,
        password: this.password,
      })
      .pipe(
        switchMap(() =>
          this.activatedRoute.queryParamMap.pipe(
            map((queryParamMap) => queryParamMap.get('redirectUrl')),
            map((redirectUrl: string) => {
              if (redirectUrl) {
                this.router.navigate([redirectUrl]);
              } else {
                this.router.navigate(['']);
              }
            })
          )
        ),
        catchError(() => {
          this.modalService.openNotifyModal({
            content: 'Incorrect email or password',
          });

          return of('Incorrect email or password');
        })
      )
      .subscribe();
  }
}
