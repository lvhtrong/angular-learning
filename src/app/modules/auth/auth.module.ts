import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SharedModule } from '@shared/shared.module';

import { AuthRoutingModule } from './auth-routing.module';
import * as fromPages from './pages';

@NgModule({
  declarations: [...fromPages.pages],
  imports: [CommonModule, AuthRoutingModule, SharedModule],
})
export class AuthModule {}
