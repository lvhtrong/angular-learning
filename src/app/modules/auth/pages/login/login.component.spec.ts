import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { AuthService } from '@core/domains/auth/services/auth.service';
import { ModalService } from '@core/services/modal/modal.service';
import { Observable, throwError } from 'rxjs';
import { MockService } from 'ng-mocks';
import { SpectatorOverrides, byText, byTestId } from '@ngneat/spectator';
import { ActivatedRoute, Router } from '@angular/router';
import { SharedModule } from '@shared/shared.module';
import { LoginComponent } from './login.component';

//#region utils

const createComponent = createComponentFactory({
  component: LoginComponent,
  providers: [
    {
      provide: AuthService,
      useValue: MockService(AuthService),
    },
    {
      provide: ActivatedRoute,
      useValue: MockService(ActivatedRoute),
    },
    {
      provide: Router,
      useValue: MockService(Router),
    },
    {
      provide: ModalService,
      useValue: MockService(ModalService),
    },
  ],
  imports: [SharedModule],
});

const setup = (options?: SpectatorOverrides<LoginComponent>) => {
  return createComponent(options);
};

const typeInField = (
  fieldName: string,
  value = 'any',
  spectator: Spectator<LoginComponent>
) => {
  const fieldElement = spectator.query(byTestId(`login-${fieldName}-field`));
  spectator.typeInElement(value, fieldElement);
};

const clickLoginButton = (spectator: Spectator<LoginComponent>) => {
  const loginButton = spectator.query(byTestId('login-login-button'));
  spectator.click(loginButton);
};

//#endregion

it('should render as expected', () => {
  const spectator = setup();

  expect(spectator.element).toMatchSnapshot();
});

describe.each([['username'], ['password']])('%s field', (field) => {
  const errorMessage = {
    username: 'User Name is required',
    password: 'Password is required',
  };

  describe('without typing then click login', () => {
    it('should show error message', () => {
      const spectator = setup();

      clickLoginButton(spectator);

      const errorElement = spectator.query(byText(errorMessage[field]));
      expect(errorElement).toBeVisible();
    });
  });

  describe('type then click login', () => {
    it('should not show error message', () => {
      const spectator = setup();

      typeInField(field, undefined, spectator);
      clickLoginButton(spectator);

      const errorElement = spectator.query(byText(errorMessage[field]));
      expect(errorElement).not.toBeVisible();
    });
  });
});

describe('type mandatory fields then click Login button', () => {
  const setupValuesThenLogin = (
    values: {
      username: string;
      password: string;
    },
    options: SpectatorOverrides<LoginComponent> = { providers: [] }
  ) => {
    const spectator = setup(options);

    typeInField('username', values.username, spectator);
    typeInField('password', values.password, spectator);
    clickLoginButton(spectator);

    return spectator;
  };

  describe('typed values are valid', () => {
    const authService = {
      login: jest
        .fn()
        .mockReturnValue(new Observable((observe) => observe.next())),
    };
    const router = {
      navigate: jest.fn(),
    };
    const modalService = {
      openNotifyModal: jest.fn(),
    };

    const username = 'username@mail.com';
    const password = '123456';

    const setupTypeValidValues = (
      options: SpectatorOverrides<LoginComponent> = { providers: [] }
    ) => {
      const spectator = setupValuesThenLogin(
        {
          username,
          password,
        },
        {
          ...options,
          providers: [
            ...options?.providers,
            {
              provide: AuthService,
              useValue: authService,
            },
            {
              provide: Router,
              useValue: router,
            },
            {
              provide: ModalService,
              useValue: modalService,
            },
          ],
        }
      );

      return spectator;
    };

    beforeEach(() => {
      modalService.openNotifyModal.mockClear();
      authService.login.mockClear();
      router.navigate.mockClear();
    });

    describe('page is opened without redirect url', () => {
      const setupPageOpenedWithoutRedirectUrl = (
        options: SpectatorOverrides<LoginComponent> = { providers: [] }
      ) => {
        const activatedRoute = {
          queryParamMap: new Observable((observe) => {
            observe.next({
              get: jest.fn().mockReturnValue(null),
            });
          }),
        };

        const spectator = setupTypeValidValues({
          ...options,
          providers: [
            ...options?.providers,
            {
              provide: ActivatedRoute,
              useValue: activatedRoute,
            },
          ],
        });

        return spectator;
      };

      it('should navigate to root url', (done) => {
        setupPageOpenedWithoutRedirectUrl();

        expect(authService.login).toHaveBeenCalledTimes(1);
        expect(authService.login).toHaveBeenCalledWith({
          username,
          password,
        });

        authService.login.mock.results[0].value.subscribe(() => {
          expect(router.navigate).toHaveBeenCalledTimes(1);
          expect(router.navigate).toHaveBeenCalledWith(['']);

          done();
        });
      });

      it('should not open error modal', (done) => {
        setupPageOpenedWithoutRedirectUrl();

        authService.login.mock.results[0].value.subscribe(() => {
          expect(modalService.openNotifyModal).not.toHaveBeenCalled();

          done();
        });
      });
    });

    describe('page is opened with redirect url', () => {
      const redirectUrl = 'redirectUrl';

      const setupPageOpenedWithRedirectUrl = (
        options: SpectatorOverrides<LoginComponent> = { providers: [] }
      ) => {
        const activatedRoute = {
          queryParamMap: new Observable((observe) => {
            observe.next({
              get: jest.fn().mockReturnValue(redirectUrl),
            });
          }),
        };

        const spectator = setupTypeValidValues({
          ...options,
          providers: [
            ...options?.providers,
            {
              provide: ActivatedRoute,
              useValue: activatedRoute,
            },
          ],
        });

        return spectator;
      };

      it('should navigate to redirect url', (done) => {
        setupPageOpenedWithRedirectUrl();

        expect(authService.login).toHaveBeenCalledTimes(1);
        expect(authService.login).toHaveBeenCalledWith({
          username,
          password,
        });

        authService.login.mock.results[0].value.subscribe(() => {
          expect(router.navigate).toHaveBeenCalledTimes(1);
          expect(router.navigate).toHaveBeenCalledWith([redirectUrl]);

          done();
        });
      });

      it('should not open error modal', (done) => {
        setupPageOpenedWithRedirectUrl();

        authService.login.mock.results[0].value.subscribe(() => {
          expect(modalService.openNotifyModal).not.toHaveBeenCalled();

          done();
        });
      });
    });
  });

  describe('typed values are invalid', () => {
    const authService = {
      login: jest.fn().mockReturnValue(
        throwError({
          error: {
            message: 'invalid username or password',
          },
        })
      ),
    };
    const router = {
      navigate: jest.fn(),
    };
    const modalService = {
      openNotifyModal: jest.fn(),
    };
    const activatedRoute = {
      queryParamMap: new Observable((observe) => {
        observe.next({
          get: jest.fn().mockReturnValue('redirectUrl'),
        });
      }),
    };

    const username = 'username@mail.com';
    const password = '123456';

    const setupTypeInvalidValues = (
      options: SpectatorOverrides<LoginComponent> = { providers: [] }
    ) => {
      const spectator = setupValuesThenLogin(
        {
          username,
          password,
        },
        {
          ...options,
          providers: [
            ...options?.providers,
            {
              provide: AuthService,
              useValue: authService,
            },
            {
              provide: Router,
              useValue: router,
            },
            {
              provide: ModalService,
              useValue: modalService,
            },
            {
              provide: ActivatedRoute,
              useValue: activatedRoute,
            },
          ],
        }
      );

      return spectator;
    };

    it('should show "Incorrect email or password" modal', (done) => {
      setupTypeInvalidValues();

      authService.login.mock.results[0].value.subscribe({
        error: () => {
          expect(modalService.openNotifyModal).toHaveBeenCalledTimes(1);
          expect(modalService.openNotifyModal).toHaveBeenCalledWith({
            content: 'Incorrect email or password',
          });

          done();
        },
      });
    });

    it('should throw error', (done) => {
      setupTypeInvalidValues();

      authService.login.mock.results[0].value.subscribe({
        error: (error: any) => {
          expect(error).toStrictEqual({
            error: {
              message: 'invalid username or password',
            },
          });

          done();
        },
      });
    });
  });
});
