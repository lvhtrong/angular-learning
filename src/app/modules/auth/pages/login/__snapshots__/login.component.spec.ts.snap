// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as expected 1`] = `
<div
  id="root0"
  ng-version="9.1.7"
>
  <div
    class="container"
  >
    <div
      class="row flex justify-content-center align-items-center h-100"
    >
      <div
        class="col-md-6 card"
      >
        <div
          class="card-body"
        >
          <h1
            class="mb-3"
          >
            Login
          </h1>
          <app-f-input-field
            id="login-username-input"
            label="Username"
            ng-reflect-id="login-username-input"
            ng-reflect-label="Username"
            ng-reflect-test-id="login-username-field"
            ng-reflect-type="email"
            ng-reflect-value=""
            testid="login-username-field"
            type="email"
          >
            <div
              class="form-group"
            >
              <label
                for="login-username-input"
              >
                Username
              </label>
              <input
                class="form-control"
                data-testid="login-username-field"
                id="login-username-input"
                type="email"
              />
              <div
                class="invalid-feedback"
              >
                  
              </div>
            </div>
          </app-f-input-field>
          <app-f-input-field
            id="login-password-input"
            label="Password"
            ng-reflect-id="login-password-input"
            ng-reflect-label="Password"
            ng-reflect-test-id="login-password-field"
            ng-reflect-type="password"
            ng-reflect-value=""
            testid="login-password-field"
            type="password"
          >
            <div
              class="form-group"
            >
              <label
                for="login-password-input"
              >
                Password
              </label>
              <input
                class="form-control"
                data-testid="login-password-field"
                id="login-password-input"
                type="password"
              />
              <div
                class="invalid-feedback"
              >
                  
              </div>
            </div>
          </app-f-input-field>
          <button
            class="btn btn-primary"
            data-testid="login-login-button"
            type="submit"
          >
             Login 
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
